### PR TITLE
Encryption support

### PR DESCRIPTION
**Pull request by GPrabhudas**
- encryption & decryption support

### CODE REVIEW

File: pulsar/internal/commands.go [hunk 1]
```diff
@@ -215,6 +216,80 @@ func addSingleMessageToBatch(wb Buffer, smm *pb.SingleMessageMetadata, payload [
...
+) error {
+	// Wire format
+	// [TOTAL_SIZE] [CMD_SIZE][CMD] [MAGIC_NUMBER][CHECKSUM] [METADATA_SIZE][METADATA] [PAYLOAD]
+
+	// compress the payload
+	compressedPayload := compressionProvider.Compress(nil, uncompressedPayload.ReadableSlice())
+
+	// encrypt payload
+	if dataKeyCrypto != nil {
+		encryptedPayload, err := msgCrypto.EncryptWithDataKeyCrypto(encryptionKeys, dataKeyCrypto, msgMetadata, compressedPayload)
```
Comment (samuelhewitt) [inline pulsar/internal/commands.go:237]: https://github.com/apache/pulsar/wiki/PIP-4:-Pulsar-End-to-End-Encryption#producer-1

Failed encryption should respect config for ProducerCryptoFailureAction

File: pulsar/consumer_partition.go [hunk 2]
```diff
@@ -478,13 +482,36 @@ func (pc *partitionConsumer) MessageReceived(response *pb.CommandMessage, header
 		return err
 	}
 
-	uncompressedHeadersAndPayload, err := pc.Decompress(msgMeta, headersAndPayload)
+	var uncompressedHeadersAndPayload internal.Buffer
+
+	// decrypt the data if needed
+	if msgMeta.EncryptionParam != nil {
+		if pc.options.messageCrypto != nil && (pc.options.cryptoKeyReader != nil || pc.options.dataKeyCrypto != nil) {
+			if pc.options.dataKeyCrypto != nil {
```
Comment (samuelhewitt) [inline pulsar/consumer_partition.go:490]: https://github.com/apache/pulsar/wiki/PIP-4:-Pulsar-End-to-End-Encryption#consumer-1

Should respect config for ConsumerCryptoFailureAction, with proper handling for when the message contains a batch of messages.

File: pulsar/consumer_partition.go [hunk 3]
```diff
@@ -478,13 +482,36 @@ func (pc *partitionConsumer) MessageReceived(response *pb.CommandMessage, header
 		return err
 	}
 
-	uncompressedHeadersAndPayload, err := pc.Decompress(msgMeta, headersAndPayload)
+	var uncompressedHeadersAndPayload internal.Buffer
+
+	// decrypt the data if needed
+	if msgMeta.EncryptionParam != nil {
```
Comment (samuelhewitt) [inline pulsar/consumer_partition.go:488]: Should be a path to deliver non-decrypted messages to client per item 2 when decryption key is not available - https://github.com/apache/pulsar/wiki/PIP-4:-Pulsar-End-to-End-Encryption#consumer-1

File: pulsar/crypto/DefaultMessageCrypto.go [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+	// create gcm param
+	nonce := make([]byte, gcm.NonceSize())
+	_, err = rand.Read(nonce)
+
+	if err != nil {
+		d.logger.Error("Failed to generate new nonce ")
+	}
+
+	// Update message metadata with encryption param
+	msgMetadata.EncryptionParam = nonce
```
Comment (samuelhewitt) [inline pulsar/crypto/DefaultMessageCrypto.go:199]: EncryptionAlgo is not set, in the java client it is also not set. Something to bring up with apache/pulsar team.

**Comment by GPrabhudas** [comment]
Comments will be taken care of next PR :)

**Pull request closed without merging**